overlord/configstate/configcore/certs.go

```diff
@@ -30,8 +31,219 @@ import (
+	}
+	return certsObjects, nil
+}
+
+// filterCerts Filters out the certificates that are not allowed to be merged.
+func filterCerts(certs []certificate, blockedCerts []string) ([]certificate, error) {
+	var filteredCerts []certificate
+	for _, cert := range certs {
+		if isBlocked(cert, blockedCerts) {
+			continue
+		}
+		filteredCerts = append(filteredCerts, cert)
+	}
+	return filteredCerts, nil
+}
+
+func filterCertsInDirectory(dirPath string, blockedCerts []string) error {
+	existingCerts, err := getCertObjects(dirPath)
+	if err != nil {
+		return err
+	}
+
+	// Remove the certs that are marked blocked.
+	for _, cert := range existingCerts {
+		if isBlocked(cert, blockedCerts) {
+			err := os.Remove(cert.Path)
+			if err != nil {
+				return fmt.Errorf("cannot remove blocked certificate: %v", err)
+			}
+		}
+	}
+	return nil
+}
+
+// installCerts Populates symbolic links in the output directory, to each certificate
+// provided.
+func installCerts(outputPath string, certs []certificate) error {
+	for _, cert := range certs {
+		// When updating existing certifications we want to make sure we don't
+		// overwrite the certification with itself.
+		certDestinationPath := filepath.Join(outputPath, cert.Name)
+		if certDestinationPath == cert.RealPath {
+			continue
+		}
+
+		// If the path exists, then remove it first to update the certificate
+		if _, err := os.Lstat(certDestinationPath); err == nil {
+			err := os.Remove(certDestinationPath)
+			if err != nil {
+				return fmt.Errorf("cannot remove existing certificate: %v", err)
+			}
+		}
+
+		// Create a link to the real path of the cert in the output path
+		err := os.Symlink(cert.RealPath, certDestinationPath)
+		if err != nil {
+			return fmt.Errorf("cannot symlink store certificate: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// generateCACertificates Generate the ca-certificates.crt to the output path
+// The ca-certificates.crt is a concatenation of all the certs in the
+// output path.
+func generateCACertificates(outputPath string) error {
+	certificates, err := getCertObjects(outputPath)
+	if err != nil {
+		return err
+	}
+
+	certsPath := filepath.Join(outputPath, "ca-certificates.crt")
+	certsFile, err := os.Create(certsPath)
+	if err != nil {
+		return fmt.Errorf("cannot create ca-certificates.crt: %v", err)
+	}
+
+	for _, cert := range certificates {
+		certBytes, err := ioutil.ReadFile(cert.RealPath)
+		if err != nil {
+			return fmt.Errorf("cannot read certificate %q: %v", cert.Name, err)
+		}
+
+		_, err = certsFile.Write(certBytes)
+		if err != nil {
+			return fmt.Errorf("cannot write certificate %q: %v", cert.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func getCertObjectsFromPaths(outputPath string, inputPaths []string) ([]certificate, error) {
+	var certObjects []certificate
+	for _, certDirectoryPath := range inputPaths {
+		// If the input directory equals output, then ignore it
+		certFullPath, err := filepath.Abs(certDirectoryPath)
```

Comment:
will we need this bit of logic in practice? I would expect in our use case output would be distinct from input?

Reply:
I'm not sure, but if users/developers will have access to this configuration on which where they may select directories they want merged, I just thought it may be plausible they would provide the output directory as an input if they wanted to keep the certs in the output (which it already does). I just wanted to take a defensive stance on unexpected input, also because the configuration interface is not 100% agreed upon yet.